proxy/src/transparency/tcp.rs

```diff
@@ -71,13 +71,32 @@ impl Proxy {
         let fut = connect.connect()
             .map_err(|e| debug!("tcp connect error: {:?}", e))
             .and_then(move |tcp_out| {
+                use futures::future::Either;
+
                 let (in_r, in_w) = tcp_in.split();
                 let (out_r, out_w) = tcp_out.split();
 
                 copy(in_r, out_w)
-                    .join(copy(out_r, in_w))
-                    .map(|_| ())
-                    .map_err(|e| debug!("tcp error: {}", e))
+                    .select2(copy(out_r, in_w))
+                    .map(|_ok| {
+                        // at this point, 1 of the copies has finished
+                        // (seen eof on its reader), and so one side
+                        // has hung up. time to go!
+                        //
+                        // TODO: we could actually try to see if the other
+                        // 'copy' has some bytes it still wants to write,
+                        // and try to write them, but we cannot do that with
+                        // the current `Copy` API. We probably just want to
+                        // implement our own duplex pipe future instead.
+                        //
+                        // However, this case is very unlikely. It'd require
+                        // that one side has closed only their write-side of
+                        // the socket, and is still waiting to read.
```

Comment:
This doesn't seem that unlikely to me, although I do think it is not super common. `shutdown(SHUT_WR)` allows the operating system to discard/recycle the send buffer and other resources on the side calling `shutdown(SHUT_WR)`; similarly the peer can discard its receive buffer. In the case of (Linux) kernel TLS the encryption (respectively, decryption) key can also be discarded. I don't know how common it is but I don't think we should speculate on that here, so I'd at least remove the last paragraph of the comment.

Regardless, I don't understand why, if we know that `copy()` doesn't do exactly what we want, why are we using `copy()`? Is it difficult to implement the correct thing?

Reply:
I originally used `copy` since I didn't think of the half-closed case. Writing a solution isn't overly difficult, but it would require essentially copying the source of the `Copy` future, and augmenting it to be duplex. My gut said this case isn't too common, and so weighed the time to write and test versus fixing this now and filing a new bug.